### PR TITLE
feat(controller): sortDAGTasks supports sort by field Depends

### PR DESCRIPTION
**Pull request by book987**
Signed-off-by: BOOK <book78987book@gmail.com>

#6306 

Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).

Tips:

* Your PR needs to pass the required checks before it can be approved. If the check is not required (e.g. E2E tests) it does not need to pass
* Sign-off your commits to pass the DCO check: `git commit --signoff`.
* Run `make pre-commit -B` to fix codegen or lint problems. 
* Say how how you tested your changes. If you changed the UI, attach screenshots.

### CODE REVIEW

File: workflow/validate/validate.go [hunk 1]
```diff
@@ -1350,14 +1348,20 @@ func verifyNoCycles(tmpl *wfv1.Template, ctx *dagValidationContext) error {
...
 
-func sortDAGTasks(tmpl *wfv1.Template) error {
+func sortDAGTasks(tmpl *wfv1.Template, ctx *dagValidationContext) error {
 	taskMap := make(map[string]*wfv1.DAGTask, len(tmpl.DAG.Tasks))
 	sortingGraph := make([]*sorting.TopologicalSortingNode, len(tmpl.DAG.Tasks))
 	for index := range tmpl.DAG.Tasks {
-		taskMap[tmpl.DAG.Tasks[index].Name] = &tmpl.DAG.Tasks[index]
+		task := tmpl.DAG.Tasks[index]
+		taskMap[task.Name] = &task
+		dependenciesMap, _ := common.GetTaskDependencies(&task, ctx)
```
Comment (sarabala1979) [inline workflow/validate/validate.go:1357]: Will common.GetTaskDependencies return sorted taskname as key?

Comment (sarabala1979) [reply workflow/validate/validate.go:1357]: https://stackoverflow.com/questions/18342784/how-to-iterate-through-a-map-in-golang-in-order

Comment (book987) [reply workflow/validate/validate.go:1357]: nope, but do we need an ordered dependencies list here? 🤔

Comment (sarabala1979) [reply workflow/validate/validate.go:1357]: My understanding is Dependencies are not in order so validation failed.  Can you test different usecases? like
a->b->c
a-> bc ->d
a->b(has global var),c ->  D depends on (B or C)

Comment (book987) [reply workflow/validate/validate.go:1357]: The reason why validation failed is `sortDAGTasks` only support sort by field `dependencies`, thus some DAG using field `depends` may failed. Here we get task's dependencies list for topological sorting only, and the order after topological sorting is the final order of DAG tasks. So I think the order of `common.GetTaskDependencies` returns is not important here.

**Review by sarabala1979 — APPROVED** [review]

**Pull request merged**